chain/client/src/sync/header.rs

```diff
@@ -237,75 +234,51 @@ impl HeaderSync {
         None
     }
 
+    // The remote side will return MAX_BLOCK_HEADERS headers, starting from the first hash in
+    // the returned "locator" list that is on their canonical chain.
+    //
+    // The locator allows us to start syncing from a reasonably recent common ancestor. Since
+    // we don't know which fork the remote side is on, we include a few hashes. The first one
+    // we include is the tip of our chain, and the next one is 2 blocks back, then 4 blocks
+    // back, then 8 blocks back, etc, until we reach the most recent final block. The reason
+    // why we stop at the final block is because the consensus guarantees us that the final
+    // blocks observed by all nodes are on the same fork.
     fn get_locator(&mut self, chain: &Chain) -> Result<Vec<CryptoHash>, near_chain::Error> {
-        let tip = chain.header_head()?;
-        let genesis_height = chain.genesis().height();
-        let heights = get_locator_heights(tip.height - genesis_height)
-            .into_iter()
-            .map(|h| h + genesis_height)
-            .collect::<Vec<_>>();
-
-        // For each height we need, we either check if something is close enough from last locator, or go to the db.
-        let mut locator: Vec<(u64, CryptoHash)> = vec![(tip.height, tip.last_block_hash)];
-        for h in heights {
-            if let Some(x) = close_enough(&self.history_locator, h) {
-                locator.push(x);
-            } else {
-                // Walk backwards to find last known hash.
-                let last_loc = *locator.last().unwrap();
-                if let Ok(header) = chain.get_block_header_by_height(h) {
-                    if header.height() != last_loc.0 {
-                        locator.push((header.height(), *header.hash()));
-                    }
-                }
-            }
+        let store = chain.store();
+        let tip = store.header_head()?;
+        // We could just get the ordinal from the header, but it's off by one: #8177.
+        let tip_ordinal = store.get_block_merkle_tree(&tip.last_block_hash)?.size();
```

Comment:
Technically, is it true that the algorithm won't work if there are more than 1000 blocks between the last final block and the most recent block? I know this is very unlikely to happen to have such long forks, but maybe here we should limit the step size to 500?

Reply:
There needs to be 1000 blocks in the canonical chain between the final head and the most recent block. So 1000 ordinals, not heights. Also, it's more like 5000 ordinals, not 1000. I actually already limiting the step size to 512 below; we'll take exponentially increasing steps until 512 per step, for up to 20 steps. Does that address your concern?